File: src/translator/zh2en/factory.ts
import { BaseTypeProvider } from "./provider/base_type.js";
import { RequirementProvider } from "./provider/requirement.js";
import { PropertyProvider } from "./provider/property.js";
import { SkillProvider } from "./provider/skill.js";
import { PassiveSkillProvider } from "./provider/passive_skill.js";
import { StatProvider } from "./provider/stat.js";
import { AttributeProvider } from "./provider/attribute.js";
import { BasicTranslator } from "./basic.js";
import { JsonTranslator } from "./json.js";
import { TextTranslator } from "./text.js";
import { StringProvider } from "./provider/sting.js";

export class TranslatorFactory {
    private basicTranslator: BasicTranslator;

    constructor() {
        const attributeProvider = new AttributeProvider();
        const baseTypeProvider = new BaseTypeProvider();
        const requirementProvider = new RequirementProvider();
        const propertyProvider = new PropertyProvider();
        const passiveSkillProvider = new PassiveSkillProvider();
        const skillProvider = new SkillProvider();
        const statProvider = new StatProvider();
        const stringProvider = new StringProvider();

        this.basicTranslator = new BasicTranslator(
            attributeProvider,
            baseTypeProvider,
            passiveSkillProvider,
            propertyProvider,
            requirementProvider,
            skillProvider,
            statProvider,
            stringProvider,
        );
    }

    getBasicTranslator(): BasicTranslator {
        return this.basicTranslator;
    }
    getJsonTranslator(): JsonTranslator {
        return new JsonTranslator(this.basicTranslator);
    }
    getTextTranslator(): TextTranslator {
        return new TextTranslator(this.basicTranslator);
    }
}

File: go/translator/zh2en/factory.go
package zh2en

import (
	"github.com/cn-poe-community/cn-poe-utils/go/translator/zh2en/provider"
)

type TranslatorFactory struct {
	basicTranslator *BasicTranslator
	jsonTranslator  *JsonTranslator
}

func NewTranslatorFactory() *TranslatorFactory {
	attributeProvider := provider.NewAttributeProvider()
	baseTypeProvider := provider.NewBaseTypeProvider()
	requirementProvider := provider.NewRequirementProvider()
	propertyProvider := provider.NewPropertyProvider()
	passiveSkillProvider := provider.NewPassiveSkillProvider()
	skillProvider := provider.NewSkillProvider()
	statProvider := provider.NewStatProvider()
	stringProvider := provider.NewStringProvider()

	basicTranslator := NewBasicTranslator(
		attributeProvider,
		baseTypeProvider,
		passiveSkillProvider,
		propertyProvider,
		requirementProvider,
		skillProvider,
		statProvider,
		stringProvider,
	)
	jsonTranslator := NewJsonTranslator(basicTranslator)

	return &TranslatorFactory{
		basicTranslator: basicTranslator,
		jsonTranslator:  jsonTranslator,
	}
}

func (f *TranslatorFactory) GetBasicTranslator() *BasicTranslator {
	return f.basicTranslator
}

func (f *TranslatorFactory) GetJsonTranslator() *JsonTranslator {
	return f.jsonTranslator
}
